File: chocofarm/solvers/ismcts.py
#!/usr/bin/env python3
"""
chocofarm policy — Single-Observer Information Set Monte-Carlo Tree Search (SO-ISMCTS).

Faithful implementation of the single-observer variant of Cowling, Powley & Whitehouse,
"Information Set Monte Carlo Tree Search", IEEE TCIAIG 4(2), 2012 (Algorithm 1, §IV-E),
with the subset-armed-bandit UCB modification of §IV-B.

Why ISMCTS fits this problem. The chocofarm task is a single-observer belief-MDP: there is
no adversary, only an "environment" that resolves each action's observation (a treasure is
present/absent; a disjunctive detector reads positive/negative). The information set the
searching player occupies is *exactly* the belief `bw` — the numpy set of latent worlds still
consistent with everything observed. So the central ISMCTS idea maps directly:

  - A node in the tree is an information set (a belief), reached by a sequence of actions whose
    *observations* came out a particular way. Edges are actions.
  - Each iteration DETERMINIZES: sample one concrete world `w ~ bw`. Because the belief already
    encodes consistency (exactly-5-of-20, all past observations applied), sampling from `bw` is
    an unbiased draw of "which world am I really in" — no particle filter needed (env note).
  - The determinization `w` RESOLVES every action's observation outcome, hence which child you
    descend into. Two visits to the same node under different determinizations branch to
    different children of the same action edge — and the action statistics aggregate over the
    whole information set, which is the property determinized-UCT lacks and ISMCTS restores.

Subset-armed bandit (§IV-B). The set of legal actions at a node SHRINKS as the belief sharpens:
a detector drops out once its outcome is certain under `bw` (env.legal_actions enforces this),
and a treasure drops out once revealed absent. So an action is "available" only on some
iterations. UCB1 (eq. 7) is therefore computed with the parent's *availability* count n'(parent)
in the log term — incremented for every action that was legal on a given visit, whether or not
it was the one selected — rather than the raw parent visit count. Without this, rarely-legal
actions are over-explored. c defaults to 0.7, the value the paper used across all experiments.

Objective coupling. The playout return is the λ-penalized renewal reward Σ value − λ·(Σ travel +
exit), i.e. the same Dinkelbach surrogate `env.dinkelbach_rate` drives the whole harness with.
At a node the agent may also choose to TERMINATE (stop and exit now); we treat TERMINATE as an
ordinary edge whose value is the bare −λ·exit_cost continuation, so the search can *learn* the
early-exit option that the clairvoyant ceiling shows is where most of the +70% lives.
"""
import math
from dataclasses import dataclass
import numpy as np
from chocofarm.solvers.base import Policy, _base_value, UCB_C, GreedyStopBase
from chocofarm.model.env import (
    Action, Collected, Environment, Loc, TERMINATE, WorldSet, is_terminate,
)

# The information-set node identity (_belief_key): a (count, first, last) fingerprint, and
# a child edge key pairs the action taken with the successor belief's fingerprint.
BeliefKey = tuple[int, int, int]
ChildKey = tuple[Action, BeliefKey]


@dataclass(frozen=True)
class ISMCTSConfig:
    """Frozen scalar hyperparameters for `ISMCTSPolicy` (audit item I). The simulation `base`
    (a Policy, not a scalar) stays a separate __init__ param. Defaults match
    `ISMCTSPolicy.__init__` so a config built from the defaults is behaviour-identical. The config
    is the single coercion home: `__post_init__` applies the same `int()`/`float()` the old
    `__init__` did, so a config-built and a kwargs-built policy hold identical typed fields."""
    iterations: int = 300
    c: float = UCB_C
    max_depth: int = 24

    def __post_init__(self) -> None:
        object.__setattr__(self, "iterations", int(self.iterations))
        object.__setattr__(self, "c", float(self.c))
        object.__setattr__(self, "max_depth", int(self.max_depth))


def _belief_key(bw: WorldSet) -> BeliefKey:
    """A cheap, order-insensitive identity for an information set (a belief world-set).

    Beliefs reached by the same observations are the same set of worlds regardless of the path;
    the smallest + largest + count triple is a collision-resistant fingerprint for the modest
    number of distinct beliefs a single search reaches (full equality is verified on collision).
    """
    n = len(bw)
    if n == 0:
        return (0, 0, 0)
    return (n, int(bw[0]), int(bw[-1]))


class _Node:
    """One information-set node. Children are keyed by action; an action's observation outcome
    under the active determinization selects which *successor belief* (sub-child) we descend to.

    We keep per-action aggregate statistics (reward sum, selection count, availability count)
    aggregated over the whole information set — that is the ISMCTS contract. The observation
    outcome only routes *which child node* the simulation continues from; it does not split the
    action's bandit statistics."""
    __slots__ = ("reward", "visits", "avail", "children")

    def __init__(self) -> None:
        self.reward: dict[Action, float] = {}   # action -> summed playout return over selections
        self.visits: dict[Action, int] = {}     # action -> times this action was selected   (n_j)
        self.avail: dict[Action, int] = {}      # action -> times this action was available  (n'_j)
        self.children: dict[ChildKey, _Node] = {}   # (action, belief_key) -> child _Node


class ISMCTSPolicy(Policy):
    """Single-Observer Information Set MCTS. `iterations` determinized tree-walks per decision;
    `c` the UCB1 exploration constant (paper default 0.7). `base` is the simulation (playout)
    policy used at leaves and for the rollout to the end of the episode; `GreedyStopBase` by
    default (pass a `policies.GreedyPolicy()` to use the plainer greedy) — cheap, and the search
    supplies the contingent depth the base lacks."""

    def __init__(self, iterations: int = 300, c: float = UCB_C, base: Policy | None = None,
                 max_depth: int = 24, *, cfg: "ISMCTSConfig | None" = None) -> None:
        # cfg=ISMCTSConfig(...) supplies (iterations, c, max_depth) in one frozen object; the
        # individual kwargs remain the back-compat path and build the config when no cfg is passed
        # (ADR-0004). `base` (a Policy, not a scalar) is always a separate __init__ param. The config
        # is the single typed home (its __post_init__ does the int()/float()); the scalars decide()
        # reads are projected straight off it, so config-built and kwargs-built policies are equal.
        self.cfg = cfg if cfg is not None else ISMCTSConfig(iterations, c, max_depth)
        self.iterations = self.cfg.iterations
        self.c = self.cfg.c
        self.base = base if base is not None else GreedyStopBase()
        self.max_depth = self.cfg.max_depth

    # ---- public API ----
    def decide(self, env: Environment, loc: Loc, bw: WorldSet, collected: Collected,
               lam: float, rng: np.random.Generator | None = None) -> Action:
        # ADR-0002 fail-loud: ISMCTS is stochastic (it determinizes a world per iteration), so
        # it requires a real Generator — matches the seam's Optional-rng contract (base.py).
        assert rng is not None, "ISMCTSPolicy.decide requires a Generator (it determinizes worlds)"
        if len(bw) == 0:
            return TERMINATE
        root = _Node()
        for _ in range(self.iterations):
            w = env.sample_world(bw, rng)              # (1) determinize: one world ~ belief
            self._iterate(env, root, loc, bw, set(collected), w, lam, rng, 0)
        # (final) return the most-visited action from the root; TERMINATE if nothing was tried.
        if not root.visits:
            return TERMINATE
        return max(root.visits, key=lambda a: root.visits[a])

    # ---- one determinized iteration: selection + expansion + simulation + backprop ----
    def _iterate(self, env: Environment, node: _Node, loc: Loc, bw: WorldSet,
                 collected: Collected, world: int, lam: float,
                 rng: np.random.Generator, depth: int) -> float:
        """Recursive descent for one iteration in fixed determinization `world`. Returns the
        λ-penalized return from `node` onward, which the caller backpropagates into the edge it
        took. `bw`/`collected` are the information-set + collected-set AT this node."""
        if depth >= self.max_depth:
            return -lam * env.exit_cost(loc)

        # Actions compatible with the determinization at this node: every legal action is
        # compatible (its observation is simply resolved by `world`); TERMINATE is always legal.
        legal = env.legal_actions(loc, bw, collected)
        actions = list(legal) + [TERMINATE]

        # Bump availability for every action legal on this visit (subset-armed bandit, §IV-B).
        for a in actions:
            node.avail[a] = node.avail.get(a, 0) + 1

        # (3) Expansion: if any compatible action is untried here, expand one at random.
        untried = [a for a in actions if a not in node.visits]
        if untried:
            a = untried[int(rng.integers(len(untried)))]
            ret = self._expand_and_simulate(env, node, loc, bw, collected, a, world, lam, rng, depth)
            self._update(node, a, ret)
            return ret

        # (2) Selection: UCB1 with availability count in the exploration term.
        a = self._ucb_select(node)
        if is_terminate(a):
            ret = -lam * env.exit_cost(loc)            # stop now: only the exit toll remains
            self._update(node, a, ret)
            return ret

        r, nloc, nbw, ncoll, dt = env.apply(loc, bw, collected, a, world)
        step = r - lam * dt
        ckey = (a, _belief_key(nbw))
        child = node.children.get(ckey)
        if child is None:
            # The action edge exists, but this determinization routes to a successor belief not
            # yet seen — create that child node (still part of the same edge's statistics).
            child = _Node()
            node.children[ckey] = child
        cont = self._iterate(env, child, nloc, nbw, ncoll, world, lam, rng, depth + 1)
        ret = step + cont
        self._update(node, a, ret)
        return ret

    def _expand_and_simulate(self, env: Environment, node: _Node, loc: Loc, bw: WorldSet,
                             collected: Collected, a: Action, world: int, lam: float,
                             rng: np.random.Generator, depth: int) -> float:
        """Realise a freshly expanded action under `world`, register its successor child, then
        (4) play the base policy to the end of the episode for the leaf estimate."""
        if is_terminate(a):
            return -lam * env.exit_cost(loc)
        r, nloc, nbw, ncoll, dt = env.apply(loc, bw, collected, a, world)
        step = r - lam * dt
        ckey = (a, _belief_key(nbw))
        if ckey not in node.children:
            node.children[ckey] = _Node()
        # base playout from the post-action belief, scored as the λ-penalized return-to-go.
        cont = _base_value(env, self.base, nloc, nbw, ncoll, world, lam)
        return step + cont

    # ---- bandit + bookkeeping ----
    def _ucb_select(self, node: _Node) -> Action:
        """UCB1 (eq. 7) with the subset-armed-bandit denominator: the log term uses the action's
        own availability count n'_j (times the action was legal here), per §IV-B."""
        best_a: Action | None = None
        best_v = -math.inf
        c = self.c
        for a, n_j in node.visits.items():
            if n_j == 0:
                return a
            exploit = node.reward[a] / n_j
            navail = node.avail.get(a, n_j)
            explore = c * math.sqrt(math.log(navail) / n_j) if navail > 1 else c
            v = exploit + explore
            if v > best_v:
                best_v, best_a = v, a
        # `_ucb_select` is only called on a fully-expanded node (visits non-empty), so a
        # best is always found; ADR-0002 fail-loud rather than a None the caller can't use.
        assert best_a is not None
        return best_a

    @staticmethod
    def _update(node: _Node, a: Action, ret: float) -> None:
        node.visits[a] = node.visits.get(a, 0) + 1
        node.reward[a] = node.reward.get(a, 0.0) + ret

File: cpp/src/ismcts.cpp
// cpp/src/ismcts.cpp
// Purpose: the C++ Single-Observer ISMCTS Policy implementation (see ismcts.hpp). A faithful
//   reimplementation of chocofarm/solvers/ismcts.py against the C++ env port — the information-set
//   node with subset-armed-bandit per-action statistics, the determinized per-iteration descent
//   (selection / expansion / simulation / backprop), the subset-armed UCB1 (eq. 7), and the
//   most-visited final — behind the composable Policy seam (ADR-0012 P2/P7: behavioral parity, NOT
//   byte-identity; the env/runner core is untouched).
//
//   DRY against the shared base (ADR-0012 P1): the leaf utility (base_value), the default leaf base
//   (GreedyStopBase), the UCB constant (UCB_C), and the generic world-sampling draw (the shared
//   RngWorldSource) all live in policy.{hpp,cpp} and are REUSED — this file does NOT include nmcs.hpp
//   and re-authors no base/sampling/leaf logic. ISMCTS's own pieces are here: the belief fingerprint,
//   the node arena, the subset-armed UCB select, and the RNG-source extension.
//
//   Parity-critical detail (the same hazard the NMCS strict-`>`/first-wins cleared): both the UCB
//   select and the most-visited final iterate over the per-action statistics in INSERTION order with
//   a strict `>` first-wins tie — Python dicts preserve insertion order and `if v > best_v` /
//   `max(...)` keep the first. C++ std::map iterates in sorted-key order, so we track insertion order
//   explicitly in `visit_order` and iterate THAT, never the map's key order.
//
// Public Domain (The Unlicense).
#include "chocofarm/ismcts.hpp"

#include <algorithm>
#include <cassert>
#include <cmath>
#include <limits>

#include "chocofarm/features.hpp"  // action_to_slot / term_slot: the action<->slot bijection

namespace chocofarm {

namespace {
// The fixed slot for an action (the action<->slot bijection, mirrors action_to_slot). ISMCTS keys
// its per-action maps by slot — a faithful stand-in for the Python Action-tuple keys (the mapping is
// a bijection), with insertion order tracked separately so the tie-break stays first-wins.
[[nodiscard]] int slot_of(const Environment& env, const Action& a) { return action_to_slot(env, a); }

// Reconstruct an Action from its slot (the inverse of action_to_slot), so run_search can return the
// selected action. Slot 0..N-1 = ("t", i); N..N+nD-1 = ("d", j); N+nD = TERMINATE.
[[nodiscard]] Action action_of_slot(const Environment& env, int slot) {
    if (slot < env.N()) return Action{ActionKind::Treasure, slot};
    if (slot < env.N() + env.n_detectors()) return Action{ActionKind::Detector, slot - env.N()};
    return terminate_action();
}

// _update(node, a, ret) (mirrors ISMCTSPolicy._update): bump visits + reward for action-slot `a`.
// The FIRST insertion of `a` into visits records its INSERTION ORDER in visit_order — the order both
// the UCB select and the most-visited final iterate (the parity-critical first-wins tie source). DENSE:
// visits[a]==0 <=> `a` not yet inserted (visits only increments from the zero-init), so the former
// `visits.find(a)==end()` first-insertion test is exactly `visits[a]==0`; reward is written in lockstep
// with visits (this is their only writer), so `+= ret` from the zero-init is the former
// `reward[a]=ret` (new) / `+= ret` (seen) — bit-identical.
void update_node(ISMCTSNode& nd, int a, double ret) {
    const size_t s = static_cast<size_t>(a);
    if (nd.visits[s] == 0) nd.visit_order.push_back(a);  // first visits-insertion -> insertion order
    nd.visits[s] += 1;
    nd.reward[s] += ret;  // dense, zero-init: += is the former (new ? ret : reward[a]+ret)
}
}  // namespace

ISMCTSPolicy::ISMCTSPolicy(const ISMCTSConfig& cfg, const Policy* base)
    : cfg_(cfg), base_(base ? base : &default_base_) {}

// ---- subset-armed UCB1 selection (eq. 7) (mirrors _ucb_select) ------------------------------------
int ISMCTSPolicy::ucb_select(const ISMCTSNode& node) const {
    int best_a = -1;
    double best_v = -std::numeric_limits<double>::infinity();
    const double c = cfg_.c;
    // iterate in INSERTION order (visit_order), not the map's sorted-key order — the parity-critical
    // first-wins tie is over insertion order (mirrors Python dict iteration in _ucb_select).
    for (int a : node.visit_order) {
        const size_t s = static_cast<size_t>(a);
        int n_j = node.visits[s];
        if (n_j == 0) return a;  // an unselected-but-present arm: pick it immediately (mirrors Python)
        double exploit = node.reward[s] / static_cast<double>(n_j);
        // avail.get(a, n_j): dense avail[a]==0 <=> the key was absent in the map (avail only increments
        // from the zero-init), so `avail[a] ? avail[a] : n_j` reproduces the map's present?value:default
        // exactly (here a slot in visit_order always had its avail bumped in the same iterate, so >0).
        int navail = (node.avail[s] != 0) ? node.avail[s] : n_j;
        double explore = (navail > 1)
                             ? c * std::sqrt(std::log(static_cast<double>(navail)) /
                                             static_cast<double>(n_j))
                             : c;
        double v = exploit + explore;
        if (v > best_v) {  // strict >: first arm (insertion order) wins a tie (mirrors `if v > best_v`)
            best_v = v;
            best_a = a;
        }
    }
    // _ucb_select is only called on a fully-expanded node (visits non-empty), so a best is always
    // found; ADR-0012 P9 — an empty visit_order here is an invariant violation (a bug), assert/abort.
    assert(best_a != -1 && "ucb_select on an empty (unexpanded) node");
    return best_a;
}

// ---- one determinized iteration (mirrors _iterate) ------------------------------------------------
double ISMCTSPolicy::iterate(const Environment& env, std::vector<ISMCTSNode>& nodes, int node,
                             const Loc& loc, const Belief& bw,
                             const CollectedSet& collected, uint32_t world, double lam,
                             ISMCTSSource& src, int depth) const {
    if (depth >= cfg_.max_depth) return -lam * env.exit_cost(loc.pt);

    // Actions compatible with the determinization at this node: every legal action is compatible
    // (its observation is simply resolved by `world`); TERMINATE is always legal. The slot order is
    // legal_actions order (treasures id-order, then detectors id-order) then TERMINATE — the SAME
    // order Python builds `actions = list(legal) + [TERMINATE]`.
    std::vector<Action> actions = env.legal_actions(bw, collected);
    actions.push_back(terminate_action());
    std::vector<int> slots;
    slots.reserve(actions.size());
    for (const Action& a : actions) slots.push_back(slot_of(env, a));

    // Bump availability for every action legal on this visit (subset-armed bandit, §IV-B). This
    // touches ONLY the avail counts; the parity-critical insertion order is the VISITS-insertion order
    // (the order _update first inserts an action into `visits`), which both _ucb_select and the
    // most-visited final iterate — so visit_order is appended at the _update site below, NOT here.
    // DENSE, zero-init: `avail[a]++` is the former `avail[a] = avail.count(a) ? avail[a]+1 : 1`
    // (avail[a]==0 <=> absent, so 0->1 on first bump, then increments — bit-identical).
    for (int a : slots) nodes[node].avail[static_cast<size_t>(a)] += 1;

    // (3) Expansion: if any compatible action is untried here, expand one uniformly. DENSE:
    // visits[a]==0 <=> untried (the former `visits.find(a)==end()`).
    std::vector<int> untried;
    for (int a : slots)
        if (nodes[node].visits[static_cast<size_t>(a)] == 0) untried.push_back(a);
    if (!untried.empty()) {
        int pick = src.expand_index(static_cast<int>(untried.size()));
        int a = untried[static_cast<size_t>(pick)];
        Action act = action_of_slot(env, a);
        double ret;
        if (act.kind == ActionKind::Terminate) {
            ret = -lam * env.exit_cost(loc.pt);  // stop now: only the exit toll remains
        } else {
            Loc nloc = loc;
            Belief nbw = bw;
            CollectedSet nc = collected;
            StepResult sr = env.apply(nloc, nbw, nc, act, world);
            double step = sr.reward - lam * sr.dt;
            // register the successor child (still part of this edge's statistics), then the leaf.
            std::tuple<int, BeliefKey> ckey{a, env.belief_key(nbw)};
            if (nodes[node].children.find(ckey) == nodes[node].children.end()) {
                nodes.emplace_back(n_action_slots(env));  // dense stats sized to the slot space
                nodes[node].children[ckey] = static_cast<int>(nodes.size()) - 1;
            }
            double cont = src.leaf_value(nloc, nbw, nc, world, lam);
            ret = step + cont;
        }
        update_node(nodes[node], a, ret);  // _update(node, a, ret)
        return ret;
    }

    // (2) Selection: UCB1 with the availability count in the exploration term.
    int a = ucb_select(nodes[node]);
    Action act = action_of_slot(env, a);
    double ret;
    if (act.kind == ActionKind::Terminate) {
        ret = -lam * env.exit_cost(loc.pt);  // stop now: only the exit toll remains
    } else {
        Loc nloc = loc;
        Belief nbw = bw;
        CollectedSet nc = collected;
        StepResult sr = env.apply(nloc, nbw, nc, act, world);
        double step = sr.reward - lam * sr.dt;
        std::tuple<int, BeliefKey> ckey{a, env.belief_key(nbw)};
        auto cit = nodes[node].children.find(ckey);
        int child;
        if (cit == nodes[node].children.end()) {
            // the action edge exists, but this determinization routes to a successor belief not yet
            // seen — create that child node (still part of the same edge's statistics).
            nodes.emplace_back(n_action_slots(env));  // dense stats sized to the slot space
            child = static_cast<int>(nodes.size()) - 1;
            nodes[node].children[ckey] = child;
        } else {
            child = cit->second;
        }
        double cont = iterate(env, nodes, child, nloc, nbw, nc, world, lam, src, depth + 1);
        ret = step + cont;
    }
    update_node(nodes[node], a, ret);  // _update(node, a, ret)
    return ret;
}

// ---- the pure search core (mirrors decide's loop + final) -----------------------------------------
Action ISMCTSPolicy::run_search(const Environment& env, const Loc& loc,
                                const Belief& bw, const CollectedSet& collected,
                                double lam, ISMCTSSource& src) const {
    if (env.empty(bw)) return terminate_action();  // mirrors decide's len(bw)==0 -> TERMINATE
    std::vector<ISMCTSNode> nodes;
    nodes.emplace_back(n_action_slots(env));  // the root (index 0); dense stats sized to the slot space
    for (int i = 0; i < cfg_.iterations; ++i) {
        uint32_t w = src.sample_world(bw);  // (1) determinize: one world ~ belief
        CollectedSet coll = collected;      // _iterate mutates a fresh collected-set per iteration
        iterate(env, nodes, 0, loc, bw, coll, w, lam, src, 0);
    }
    // (final) most-visited root action; TERMINATE if nothing was tried. First-wins tie over INSERTION
    // order (mirrors max(root.visits, key=...) which keeps the first max). DENSE: "nothing was tried" is
    // `visit_order.empty()` (the dense visits vector is always size n_slots; the former `visits.empty()`
    // meant the MAP had no key, i.e. no slot was ever inserted — exactly visit_order being empty).
    const ISMCTSNode& root = nodes[0];
    if (root.visit_order.empty()) return terminate_action();
    int best_slot = -1;
    int best_visits = -1;
    for (int a : root.visit_order) {
        if (root.visits[static_cast<size_t>(a)] == 0) continue;  // an avail-only arm never selected: skip
        if (root.visits[static_cast<size_t>(a)] > best_visits) {  // strict >: first arm wins a tie
            best_visits = root.visits[static_cast<size_t>(a)];
            best_slot = a;
        }
    }
    assert(best_slot != -1 && "non-empty visits but no most-visited arm");
    return action_of_slot(env, best_slot);
}

namespace {
// The production ISMCTS source: the generic uniform sample_world (reused from the shared
// RngWorldSource), the expansion-index draw (rng.integers(n)), and the GreedyStopBase base_value
// leaf. It DERIVES from RngWorldSource so the world draw is the shared one (ADR-0012 P1); it adds the
// two ISMCTS-specific draws + the real leaf. Production wires off the real rng + the real base.
class RngISMCTSSource final : public ISMCTSSource {
  public:
    RngISMCTSSource(const Environment& env, const Policy& base, std::mt19937_64& rng)
        : env_(env), base_(base), draw_(env, rng), rng_(rng) {}  // env homes the uniform world draw (L1)

    uint32_t sample_world(const Belief& bw) override { return draw_.sample_world(bw); }

    int expand_index(int n) override {
        std::uniform_int_distribution<int> pick(0, n - 1);  // mirrors int(rng.integers(n))
        return pick(rng_);
    }

    double leaf_value(const Loc& loc, const Belief& bw,
                      const CollectedSet& collected, uint32_t world, double lam) override {
        return base_value(env_, base_, loc, bw, collected, world, lam);  // shared leaf utility (P1)
    }

  private:
    const Environment& env_;
    const Policy& base_;
    RngWorldSource draw_;   // the shared generic uniform-from-belief draw (ADR-0012 P1)
    std::mt19937_64& rng_;  // the SAME stream the draw uses, for the expansion-index draw
};
}  // namespace

Action ISMCTSPolicy::decide(const Environment& env, const Loc& loc, const Belief& bw,
                            const CollectedSet& collected, double lam,
                            std::mt19937_64& rng) const {
    RngISMCTSSource src(env, *base_, rng);
    return run_search(env, loc, bw, collected, lam, src);
}

}  // namespace chocofarm
